compiler/rustc_ast_lowering/src/lifetime_collector.rs

```diff
@@ -0,0 +1,83 @@
+use super::ResolverAstLoweringExt;
+use rustc_ast::visit::{self, BoundKind, LifetimeCtxt, Visitor};
+use rustc_ast::{
+    FnRetTy, GenericBounds, Lifetime, NodeId, PolyTraitRef, TraitBoundModifier, Ty, TyKind,
+};
+use rustc_hir::def::LifetimeRes;
+use rustc_middle::ty::ResolverAstLowering;
+use rustc_span::Span;
+
+struct LifetimeCollectVisitor<'this, 'ast: 'this> {
+    resolver: &'this ResolverAstLowering,
+    current_binders: Vec<NodeId>,
+    collected_lifetimes: Vec<&'ast Lifetime>,
+    underscore_lifetimes: Vec<(NodeId, Span)>,
+}
+
+impl<'this, 'ast: 'this> LifetimeCollectVisitor<'this, 'ast> {
+    fn new(resolver: &'this ResolverAstLowering) -> Self {
+        Self {
+            resolver,
+            current_binders: Vec::new(),
+            collected_lifetimes: Vec::new(),
+            underscore_lifetimes: Vec::new(),
+        }
+    }
+}
+
+impl<'this, 'ast: 'this> Visitor<'ast> for LifetimeCollectVisitor<'this, 'ast> {
+    fn visit_lifetime(&mut self, lifetime: &'ast Lifetime, _: LifetimeCtxt) {
+        let res = self.resolver.get_lifetime_res(lifetime.id).unwrap_or(LifetimeRes::Error);
+
+        if let Some(binder) = res.binder() && !self.current_binders.contains(&binder) {
+            if !self.collected_lifetimes.contains(&lifetime) {
+                self.collected_lifetimes.push(lifetime);
+            }
+        }
+    }
+
+    fn visit_poly_trait_ref(&mut self, t: &'ast PolyTraitRef, m: &'ast TraitBoundModifier) {
+        self.current_binders.push(t.trait_ref.ref_id);
+
+        visit::walk_poly_trait_ref(self, t, m);
+
+        self.current_binders.pop();
+    }
+
+    fn visit_ty(&mut self, t: &'ast Ty) {
+        match t.kind {
+            TyKind::BareFn(_) => {
+                self.current_binders.push(t.id);
+                visit::walk_ty(self, t);
+                self.current_binders.pop();
+            }
+            TyKind::Rptr(Some(_), _) => {
```

Comment:
```suggestion
            TyKind::Rptr(None, _) => {
```
We are interested in the case with an elided lifetime.